File: nutrition.py
from django.db import models


class FoodItem(models.Model):
    name = models.CharField(max_length=32)
    brand = models.ForeignKey("main.Brand", models.CASCADE)
    cost = models.IntegerField()
    gtin = models.CharField(max_length=14)

    serving_size = models.FloatField()
    serving_unit = models.CharField(max_length=16)
    serving_size_alt = models.CharField(
        max_length=32, help_text="Alternate human-readable serving size, e.g. '1 box'"
    )
    calories = models.IntegerField()
    nutrition = models.ManyToManyField("Nutrition")


class Nutrition(models.Model):
    field = models.ForeignKey("Nutrient", models.PROTECT)
    value = models.FloatField()


class Nutrient(models.Model):
    name = models.CharField(max_length=32)
    unit = models.CharField(max_length=16)

File: lib/Nutrition.ts
import * as Yup from 'yup';
import { ObjectId } from 'mongodb';

class FoodItem {
  public name: string;
  public brand: ObjectId;
  public cost: number;
  public gtin: string;
  public nutrition: {
    servingSize: number;
    servingUnit: string;
    servingSizeAlt: string;
    calories: number;
    nutritionInfo: ObjectId;
  };

  public static schema = Yup.object().shape({
    name: Yup.string().required(),
    brand: Yup.object(),
    cost: Yup.number(),
    gtin: Yup.string(),
    nutrition: Yup.object().shape({
      servingSize: Yup.number(),
      servingUnit: Yup.string(),
      servingSizeAlt: Yup.string(),
      calories: Yup.number(),
      nutritionInfo: Yup.object(),
    }),
  });

  constructor(obj: any) {
    this.name = obj.name;
    this.brand = obj.brand;
    this.cost = obj.cost;
    this.gtin = obj.gtin;
    this.nutrition = obj.nutrition;
  }
}
